File: app/src/main/java/xyz/b515/schedule/ui/view/MainActivity.java
package xyz.b515.schedule.ui.view;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;

import java.util.Calendar;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import java8.util.stream.Collectors;
import java8.util.stream.RefStreams;
import xyz.b515.schedule.Constant;
import xyz.b515.schedule.R;

public class MainActivity extends AppCompatActivity {
    @BindView(R.id.tab_layout) TabLayout tabLayout;
    @BindView(R.id.view_pager) ViewPager viewPager;
    @BindView(R.id.toolbar) Toolbar toolbar;
    @BindView(R.id.title) TextView title;
    @BindView(R.id.spinner) Spinner spinner;
    SharedPreferences preferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        title.setText(R.string.app_name);
        preferences = PreferenceManager.getDefaultSharedPreferences(this);
        checkCurrentWeek();

        setSupportActionBar(toolbar);

        viewPager.setAdapter(new CourseFragmentAdapter(getSupportFragmentManager()));
        tabLayout.setupWithViewPager(viewPager);

        List<String> weeks = RefStreams.iterate(1, i -> i + 1)
                .limit(20)
                .map(i -> String.format(getResources().getString(R.string.week_num), i))
                .collect(Collectors.toList());
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, R.layout.spinner_item, weeks);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);
        spinner.setSelection(preferences.getInt(Constant.CURRENT_WEEK, 0));
        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                preferences.edit().putInt(Constant.CURRENT_WEEK, position).apply();
                reload();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        reload();
    }

    private void reload() {
        int page = viewPager.getCurrentItem();
        viewPager.setAdapter(new CourseFragmentAdapter(getSupportFragmentManager()));
        viewPager.setCurrentItem(page);
    }

    private void checkCurrentWeek() {
        int newWeekOfYear = Calendar.getInstance().get(Calendar.WEEK_OF_YEAR);
        int oldWeek = preferences.getInt(Constant.CURRENT_WEEK, -1);
        int oldWeekOfYear = preferences.getInt(Constant.CURRENT_WEEK_OF_YEAR, newWeekOfYear - 1);
        if (newWeekOfYear > oldWeekOfYear) {
            int newWeek = oldWeek + newWeekOfYear - oldWeekOfYear;
            preferences.edit().putInt(Constant.CURRENT_WEEK, newWeek < 20 ? newWeek : 0)
                    .putInt(Constant.CURRENT_WEEK_OF_YEAR, newWeekOfYear)
                    .apply();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_manage: {
                Intent intent = new Intent(this, CourseManageActivity.class);
                startActivity(intent);
            }
            break;
            case R.id.action_settings: {
                Intent intent = new Intent(this, SettingsActivity.class);
                startActivity(intent);
            }
            break;
            case R.id.action_capture: {

            }
            break;
        }
        return super.onOptionsItemSelected(item);
    }

    private class CourseFragmentAdapter extends FragmentPagerAdapter {

        CourseFragmentAdapter(FragmentManager manager) {
            super(manager);
        }

        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    return new TodayCoursesFragment();
                default:
                    return new WeekCoursesFragment();
            }
        }

        @Override
        public int getCount() {
            return 2;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return getResources().getString(R.string.main_today_courses);
                default:
                    return getResources().getString(R.string.main_week_courses);
            }
        }
    }
}

File: app/src/main/kotlin/xyz/b515/schedule/ui/view/MainActivity.kt
package xyz.b515.schedule.ui.view

import android.app.FragmentManager
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v13.app.FragmentPagerAdapter
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import kotlinx.android.synthetic.main.activity_main.*
import xyz.b515.schedule.Constant
import xyz.b515.schedule.R
import java.util.*

class MainActivity : AppCompatActivity() {
    lateinit var preferences: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        title_text.setText(R.string.app_name)
        preferences = PreferenceManager.getDefaultSharedPreferences(this)
        checkCurrentWeek()

        setSupportActionBar(toolbar)

        view_pager.adapter = CourseFragmentAdapter(fragmentManager)
        tab_layout.setupWithViewPager(view_pager)

        val weeks = (1..20).map { getString(R.string.week_num).format(it) }
        val adapter = ArrayAdapter(this, R.layout.spinner_item, weeks)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = adapter
        spinner.setSelection(preferences.getInt(Constant.CURRENT_WEEK, 0))
        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View?, position: Int, id: Long) {
                preferences.edit().putInt(Constant.CURRENT_WEEK, position).apply()
                reload()
            }

            override fun onNothingSelected(parent: AdapterView<*>) = Unit
        }
    }

    override fun onResume() {
        super.onResume()
        reload()
    }

    private fun reload() {
        val page = view_pager.currentItem
        view_pager.adapter = CourseFragmentAdapter(fragmentManager)
        view_pager.currentItem = page
    }

    private fun checkCurrentWeek() {
        val newWeekOfYear = Calendar.getInstance().get(Calendar.WEEK_OF_YEAR)
        val oldWeek = preferences.getInt(Constant.CURRENT_WEEK, -1)
        val oldWeekOfYear = preferences.getInt(Constant.CURRENT_WEEK_OF_YEAR, newWeekOfYear - 1)
        if (newWeekOfYear > oldWeekOfYear) {
            val newWeek = oldWeek + newWeekOfYear - oldWeekOfYear
            preferences.edit().putInt(Constant.CURRENT_WEEK, if (newWeek < 20) newWeek else 0)
                    .putInt(Constant.CURRENT_WEEK_OF_YEAR, newWeekOfYear)
                    .apply()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_manage -> {
                val intent = Intent(this, CourseManageActivity::class.java)
                startActivity(intent)
            }
            R.id.action_settings -> {
                val intent = Intent(this, SettingsActivity::class.java)
                startActivity(intent)
            }
            R.id.action_capture -> {

            }
        }
        return super.onOptionsItemSelected(item)
    }

    private inner class CourseFragmentAdapter(manager: FragmentManager) : FragmentPagerAdapter(manager) {
        override fun getItem(position: Int) = when (position) {
            0 -> TodayCoursesFragment()
            else -> WeekCoursesFragment()
        }

        override fun getCount() = 2

        override fun getPageTitle(position: Int): String = when (position) {
            0 -> getString(R.string.main_today_courses)
            else -> getString(R.string.main_week_courses)
        }
    }
}
